### PR TITLE
network: fixes for lifetime handling

### LINKED ISSUE #23467: networkd: failed to set ipv6 address to interface
**systemd version the issue has been seen with**
 systemd 250 (250.5-1-arch)

**Used distribution**
ArchLinux

**Linux kernel version used** (`uname -a`)

`Linux m3 5.17.9-arch1-1 #1 SMP PREEMPT Wed, 18 May 2022 17:30:11 +0000 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
x86-64

**Expected behaviour you didn't see**
* IPv6 address is successfully set
* No error logged
* the state of `br0` interface is `(configured)`.

**Unexpected behaviour you saw**
The state of `br0` is `routable(failed)` and IPv6 address is not set correctly.

```
# networkctl status br0    
● 3: br0                                                                   
                     Link File: /usr/lib/systemd/network/99-default.link
                  Network File: /etc/systemd/network/br0.network
                          Type: bridge
                         State: routable (failed)
                  Online state: online                                     
                        Driver: bridge
              Hardware Address: ca:91:c4:31:2b:c3
                           MTU: 1500 (min: 68, max: 65535)
                         QDisc: noqueue
  IPv6 Address Generation Mode: eui64
                 Forward Delay: 14.990s
                    Hello Time: 1.990s
                       Max Age: 19.990s
                   Ageing Time: 4min 59.990s
                      Priority: 32768
                           STP: no
        Multicast IGMP Version: 2
                          Cost: 1999
                    Port State: disabled
          Queue Length (Tx/Rx): 1/1
              Auto negotiation: no
                         Speed: 1Gbps
                       Address: 192.168.2.5 (DHCP4 via 192.168.2.1)
                                fdc9:760b:4d02:0:2b64:6db8:3bc2:28c9
                                fdc9:760b:4d02:0:c891:c4ff:fe31:2bc3
                                fe80::c891:c4ff:fe31:2bc3
                       Gateway: 192.168.2.1
                           DNS: 192.168.2.10
                                fdc9:760b:4d02::1
             Activation Policy: up
           Required For Online: yes
               DHCP4 Client ID: IAID:0xa731b006/DUID
             DHCP6 Client IAID: 0xa731b006
             DHCP6 Client DUID: DUID-EN/Vendor:0000ab11f491e6d3f044088f0000

May 21 14:50:09 m3 systemd-networkd[700]: br0: Gained carrier
May 21 14:50:09 m3 systemd-networkd[700]: br0: DHCPv4 address 192.168.2.5/24, gateway 192.168.2.1 acquired from 192.168.2.1
May 21 14:50:10 m3 systemd-networkd[700]: br0: DHCPv6 address fdc9:760b:4d02::876/128 (valid for 11h 59min 59s, preferred for 36min 51s)
May 21 14:50:10 m3 systemd-networkd[700]: br0: DHCPv6 address 2x:x:x:9a0::876/128 (valid for 36min 51s, preferred for 36min 51s)
May 21 14:50:10 m3 systemd-networkd[700]: br0: DHCP: received delegated prefix fdc9:760b:4d02:4::/62
May 21 14:50:10 m3 systemd-networkd[700]: br0: DHCP: received delegated prefix 2xx:xx:xx:c154::/62
May 21 18:02:51 m3 systemd-networkd[700]: br0: DHCPv6 address 2xx:xx:xx:9a0::876/128 (valid for 59min 13s, preferred for 59min 13s)
May 21 18:02:51 m3 systemd-networkd[700]: br0: DHCP: received delegated prefix 2xx:xx:xx:9a4::/62
May 21 19:00:28 m3 systemd-networkd[700]: br0: Could not set NDisc address: Invalid argument
May 21 19:00:28 m3 systemd-networkd[700]: br0: Failed
```

**Steps to reproduce the problem**
1. boot system
2. wait DHCPv6 renew the IPv6 address.
3. error happend

**Additional program output to the terminal or log subsystem illustrating the issue**
I use a OpenWRT router, and the IPv6 dhcp mode is setted to "hybrid". So there is multiple addresses delegated. One from the ISP, and other from OpenWRT itself.

It seems the ISP's address only valid less than 1 hour. Before this address is expired, systemd-networkd try to renew the address. But it seems that systemd-networkd failed with `Invalid argument` error.

I have tried manually add the IPv6 address to `br0` using these command, and both of them are successful.
```
sudo ip addr add 2xx:xx:xx:9a0::876/128 dev br0
sudo ip addr add 2xx:xx:xx:9a4::/62 dev br0
```

And this is the failed NETLINK message (captured using strace), hope this helps:

```
sendto(3, [{nlmsg_len=80, nlmsg_type=RTM_NEWADDR, nlmsg_flags=NLM_F_REQUEST|NLM_F_ACK|NLM_F_REPLACE, nlmsg_seq=5129761, nlmsg_pid=0}, {ifa_family=AF_INET6, ifa_prefixlen=64, ifa_flags=0, ifa_scope=RT_SCOPE_UNIVERSE, ifa_index=if_nametoindex("br0")}, [[{nla_len=8, nla_type=IFA_FLAGS}, IFA_F_MANAGETEMPADDR|IFA_F_NOPREFIXROUTE], [{nla_len=20, nla_type=IFA_LOCAL}, inet_pton(AF_INET6, "2xx:xx:xx:6c80:c891:c4ff:fe31:2bc3")], [{nla_len=20, nla_type=IFA_CACHEINFO}, {ifa_prefered=0, ifa_valid=0, cstamp=0, tstamp=0}], [{nla_len=8, nla_type=IFA_RT_PRIORITY}, 0]]], 80, 0, {sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, 128) = 80

recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=128 => 12, msg_iov=[{iov_base=[{nlmsg_len=100, nlmsg_type=NLMSG_ERROR, nlmsg_flags=0, nlmsg_seq=5129761, nlmsg_pid=1}, {error=-EINVAL, msg=[{nlmsg_len=80, nlmsg_type=RTM_NEWADDR, nlmsg_flags=NLM_F_REQUEST|NLM_F_ACK|NLM_F_REPLACE, nlmsg_seq=5129761, nlmsg_pid=0}, {ifa_family=AF_INET6, ifa_prefixlen=64, ifa_flags=0, ifa_scope=RT_SCOPE_UNIVERSE, ifa_index=if_nametoindex("br0")}, [[{nla_len=8, nla_type=IFA_FLAGS}, IFA_F_MANAGETEMPADDR|IFA_F_NOPREFIXROUTE], [{nla_len=20, nla_type=IFA_LOCAL}, inet_pton(AF_INET6, "2xx:xx:xx:6c80:c891:c4ff:fe31:2bc3")], [{nla_len=20, nla_type=IFA_CACHEINFO}, {ifa_prefered=0, ifa_valid=0, cstamp=0, tstamp=0}], [{nla_len=8, nla_type=IFA_RT_PRIORITY}, 0]]]}], iov_len=2792}], msg_iovlen=1, msg_control=[{cmsg_len=20, cmsg_level=SOL_NETLINK, cmsg_type=0x3}], msg_controllen=24, msg_flags=0}, MSG_TRUNC) = 100
```

This is my network configure files on br0:
```
# /etc/systemd/network/br0.netdev
[NetDev]
Name=br0
Kind=bridge
```

```
# /etc/systemd/network/br0.network
[Match]
Name=br0

[Network]
DHCP=yes
IPv6PrivacyExtensions=yes
IPv6AcceptRA=yes
```

### LINKED ISSUE #23625: networkd reverts to link-local addresses after a few days
**systemd version the issue has been seen with**

 > systemd 251 (251.1-1-arch)

**Used distribution**
 > Arch Linux

**Linux kernel version used** (`uname -a`)

 > 5.17.9-arch1-1

**CPU architecture issue was seen on**
 > amd64

**Expected behaviour you didn't see**
 > networkd to keep my connection up

**Unexpected behaviour you saw**
 > After a few days networkd only renewed fe80:: 

**Steps to reproduce the problem**
 > Connect to a network and leave it running for a few days

**Additional program output to the terminal or log subsystem illustrating the issue**

I've pasted details of my VPS and debug logs up to the failure: https://gist.github.com/Jookia/ece8bb0ef6d71718ff3d0632781f4421
I can provide the full log but it's 464M, I would have to compress it.

The fail seems to be:

```
Jun 03 04:22:05 blane.jookia.org systemd-networkd[920]: ens3: ndisc_check_ready(): SLAAC addresses and routes are not set.
Jun 03 04:22:05 blane.jookia.org systemd-networkd[920]: ens3: ndisc_check_ready(): SLAAC addresses and routes are not set.
Jun 03 04:22:05 blane.jookia.org systemd-networkd[920]: ens3: Could not set NDisc address: Invalid argument
Jun 03 04:22:05 blane.jookia.org systemd-networkd[920]: ens3: Failed
Jun 03 04:22:05 blane.jookia.org systemd-networkd[920]: ens3: State changed: configured -> failed
Jun 03 04:22:05 blane.jookia.org systemd-networkd[920]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_32 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=21 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
```

Then it stops renewing IPv6 and IPv4 addresses.

Possibly the same issue as #23467

Edit: Unit file:

```
[Match]
Type=ether

[Network]
IPv6PrivacyExtensions=Yes
IPv6AcceptRA=yes
DHCP=yes
DNS=2001:4860:4860::8888 8.8.8.8
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #23625.
Fixes #23467.

### CODE REVIEW

**Comment by Jookia** [comment]
I've installed c6ef96c on my machine affected by the dropout. I'll update in a few days on whether this fixes #23625

File: src/network/networkd-address.c [hunk 1]
```diff
@@ -228,13 +229,13 @@ static struct ifa_cacheinfo *address_set_cinfo(const Address *a, struct ifa_cach
         return cinfo;
 }
 
-static void address_set_lifetime(Address *a, const struct ifa_cacheinfo *cinfo) {
+static void address_set_lifetime(Manager *m, Address *a, const struct ifa_cacheinfo *cinfo) {
         usec_t now_usec;
 
         assert(a);
         assert(cinfo);
```
Comment (bluca) [inline src/network/networkd-address.c:236]: `assert(m)`

File: src/network/networkd-dhcp-prefix-delegation.c [hunk 2]
```diff
@@ -869,7 +869,8 @@ static int dhcp4_pd_assign_subnet_prefix(Link *link, Link *uplink) {
         if (r < 0)
                 return log_link_warning_errno(uplink, r, "Failed to get lifetime of DHCPv4 lease: %m");
```
Comment (bluca) [inline src/network/networkd-dhcp-prefix-delegation.c:872]: `assert(link->manager)` above

File: src/network/networkd-dhcp-prefix-delegation.c [hunk 3]
```diff
@@ -941,7 +942,8 @@ int dhcp4_pd_prefix_acquired(Link *uplink) {
...
 
-        lifetime_usec = usec_add(lifetime_sec * USEC_PER_SEC, now(CLOCK_BOOTTIME));
+        assert_se(sd_event_now(link->manager->event, CLOCK_BOOTTIME, &now_usec) >= 0);
```
Comment (bluca) [inline src/network/networkd-dhcp-prefix-delegation.c]: uhm `link` is not initialised at this point? did you mean `uplink->`?

File: src/network/networkd-ndisc.c [hunk 4]
```diff
@@ -1145,20 +1145,20 @@ int link_request_ndisc(Link *link) {
 void ndisc_vacuum(Link *link) {
         NDiscRDNSS *r;
         NDiscDNSSL *d;
-        usec_t time_now;
+        usec_t now_usec;
 
         assert(link);
```
Comment (bluca) [inline src/network/networkd-ndisc.c:1152]: `assert(link->manager)`

File: src/network/networkd-dhcp4.c [hunk 5]
```diff
@@ -814,12 +814,14 @@ static int dhcp4_request_address(Link *link, bool announce) {
...
         if (!FLAGS_SET(link->network->keep_configuration, KEEP_CONFIGURATION_DHCP)) {
                 uint32_t lifetime_sec;
+                usec_t now_usec;
 
                 r = sd_dhcp_lease_get_lifetime(link->dhcp_lease, &lifetime_sec);
                 if (r < 0)
                         return log_link_warning_errno(link, r, "DHCP error: no lifetime: %m");
 
-                lifetime_usec = usec_add(lifetime_sec * USEC_PER_SEC, now(CLOCK_BOOTTIME));
+                assert_se(sd_event_now(link->manager->event, CLOCK_BOOTTIME, &now_usec) >= 0);
```
Comment (bluca) [inline src/network/networkd-dhcp4.c:824]: `assert(link->manager)`

File: src/network/networkd-dhcp-prefix-delegation.c [hunk 6]
```diff
@@ -869,8 +869,8 @@ static int dhcp4_pd_assign_subnet_prefix(Link *link, Link *uplink) {
         if (r < 0)
                 return log_link_warning_errno(uplink, r, "Failed to get lifetime of DHCPv4 lease: %m");
 
-        assert_se(sd_event_now(link->manager->event, CLOCK_BOOTTIME, &now_usec) >= 0);
-        lifetime_usec = usec_add(lifetime_sec * USEC_PER_SEC, now_usec);
+        assert_se(sd_event_now(uplink->manager->event, CLOCK_BOOTTIME, &now_usec) >= 0);
```
Comment (bluca) [inline src/network/networkd-dhcp-prefix-delegation.c:873]: this changes from link to uplink in the last commit, but doesn't seem related? same below

File: src/network/networkd-address.c (hunk 1 above)
Comment (bluca) [reply src/network/networkd-address.c:236]: looks like it was added in the last commit, but should be in the first

**Review by bluca — APPROVED** [review]

**Pull request merged**